src/Block/AdminListBlockService.php

```diff
@@ -41,23 +41,52 @@ class AdminListBlockService extends AbstractBlockService
     private $templateRegistry;
 
     /**
-     * NEXT_MAJOR: Remove `$templating` argument.
+     * NEXT_MAJOR: Remove `$deprecatedTemplatingOrNull` argument.
      *
-     * @param Environment|string $twigOrName
+     * @param Environment|string $twigOrDeprecatedName
      */
     public function __construct(
-        $twigOrName,
-        ?EngineInterface $templating,
+        $twigOrDeprecatedName,
+        ?EngineInterface $deprecatedTemplatingOrNull,
         Pool $pool,
         ?TemplateRegistryInterface $templateRegistry = null
     ) {
-        parent::__construct($twigOrName, $templating);
+        // NEXT_MAJOR: Remove conditions, only pass the Environment
+        if ($deprecatedTemplatingOrNull instanceof EngineInterface) {
+            if (!\is_string($twigOrDeprecatedName)) {
+                throw new \TypeError(sprintf(
+                    'Passing other than string as argument 1 to %s() is not allowed when %s is passed as argument 2.'
+                    .' You must pass an string instead.',
+                    __METHOD__,
+                    EngineInterface::class
+                ));
+            }
+
+            parent::__construct($twigOrDeprecatedName, $deprecatedTemplatingOrNull);
+        } elseif ($twigOrDeprecatedName instanceof Environment) {
+            if (null !== $deprecatedTemplatingOrNull) {
+                throw new \TypeError(sprintf(
+                    'Argument 2 passed to %s() must be null if you passing %s as first argument, %s given.',
+                    __METHOD__,
+                    Environment::class,
+                    EngineInterface::class
+                ));
+            }
+
+            parent::__construct($twigOrDeprecatedName);
+        } else {
+            throw new \TypeError(sprintf(
+                'Argument 1 passed to %s() should be an instance of %s, and you should leave argument 2 as null.',
+                __METHOD__,
+                Environment::class
+            ));
+        }
 
```

Comment:
Are we need this?
Parent class will trigger this:
https://github.com/sonata-project/SonataBlockBundle/blob/3.x/src/Block/Service/AbstractBlockService.php#L51-L94

Only `TypeError` should be throw if someone use BlockBundle v4 and don't use `Environment` as Argument 1.

Reply:
BTW, It can be merge and improve in other PR like you wrote in https://github.com/sonata-project/SonataAdminBundle/pull/6212#discussion_r457568984. I can do it if you don't want to do it.

Reply:
IMO we should only check if first argument is Environment and second is null, then use new construct, otherwise use the 2 parameters. And we leverage all the deprecations and errors to the BlockBundle (like I had before)

If someone pass old params to 3.x = deprecation
Old params to 4.x = type error

Otherwise good to go

Reply:
```suggestion 
if (!class_exists('AbstractAdminBlockService')) {
     parent::__constructor($twigOrDeprecatedName); // TypeError will be throw by type hints
} else {
    parent::__construct($twigOrDeprecatedName, $deprecatedTemplatingOrNull); // by parent
}
```

Reply:
@jordisala1991 WDYT about it? It is much simpler

Reply:
I will try it